### DISCUSSION TITLE
How to tell sympy to use intermediate definitions for huge expressions

### DISCUSSION THREAD

**Discussion by yota-code** [sympy/sympy#27057] [Q&A] [answered]
When using the CAS of matlab I found this behavior very nice: if an expression N is quite big, but some parts of it just repeat, the result of a simplification might be presented in more that one line, with new intermediate definitions proposed by the CAS, like this

```
>>> N.simplify() # returns
let S = <some sub expression>
N = S / (S + 1)**2
```

I hope my example is understandable enough...

Can I ask sympy to do something similar? Where, instead of explicitly telling sympy to substitute complex sub-expressions, it would propose sensitive substitutions

**Comment by moorepants** [comment, marked as answer]
SymPy has the function `cse()` that can turn any expression into a set of common subexpressions.

**Reply by yota-code** [reply]
amazing !
